### DISCUSSION TITLE
Inherit Params from separate YML file

### DISCUSSION THREAD

**Discussion by GScottAlder** [quarto-dev/quarto-cli#3658] [Q&A]
NOTE:
I am using the quarto R package version 1.0 and quarto version 0.9.80
I found the version by running this code: `as.numeric_version(system2(quarto:::find_quarto(), "--version", stdout = TRUE))`
Due to some IT restrictions I have not updated quarto to the most recent release, although I may be able to in the future.

I have multiple *.qmd files that I want to be parameterized. I want them to share parameters and default values by inheriting from a separate YML file.

For example, I have a file named test.qmd with this YML header:

```
---
title: "Test Quarto Report"
metadata-files:
  - _params.yml
---
```

Then, in _params.yml I have

```
---
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
---
```

However, this does not work. When I run quarto render, I get an error for "object 'params' not found"

Is there anything I can do to inherit params? Is this not possible in quarto 0.9.80? If so, is it possible in a later release?

Thanks,
Scott

**Comment by cscheid** [comment]
That version is a prerelease that's almost a year old now, and quarto is itself just over two years old. You really should upgrade to one of the stable versions (ideally, 1.2.269, the latest release).

**Reply by GScottAlder** [reply]
Alright, I switched to a different PC. I am now using quarto R package version 1.2 and quarto version 1.2.269

I still get the same error: object 'params' not found

I think my question boils down to: Is it possible for *.qmd files to inherit from a separate YML file? Or is that only possible for project level YML files (such as _quarto.yml)?

Thanks

**Reply by GScottAlder** [reply]
I also tried putting this into a project level _quarto.yml file

metadata-files:
  - "_params.yml"

Still doesn't work

**Reply by cscheid** [reply]
Oh, I see the problem now. Your params.yml file should not have the `---` front matter markers in them.

**Reply by GScottAlder** [reply]
Ah sorry, I did not actually have the `---` marks in the _params.yml. That was a typo in my post.

test.qmd file:
```
---
title: "Test"
format: html
metadata-files:
  - "_params.yml"
---

# Test

` ` `{r}
print(params$p1)
print(params$p2)
print(params$p3)
` ` `
```

_params.yml:
```
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
```


Does not work:
Error in print(params$p1): object 'params' not found

**Reply by cscheid** [reply]
Yeah, I'm pretty sure this is a bug, thanks for the patience.

**Reply by GScottAlder** [reply]
Okay, good to know. I wasn't sure if it was intended or not that `metadata-files:` was not working in this case.

Thanks,
Scott

**Reply by cscheid** [reply]
Tracking in #3663.

In the meantime, you might be able to use `quarto render file.qmd --execute-params params.yml`: https://quarto.org/docs/computations/parameters.html

**Reply by GScottAlder** [reply]
Yeah I can definitely use this as a workaround in the meantime.

The main downside here IMO is that I have multiple parameterized .qmd files in a project and it seems that they each need to have all the parameters listed to have the website render. So I need to maintain some duplicate YML front matter across many files. For the meantime the workaround is fine though.

Thanks

**Reply by mcanouil** [reply]
There is a workaround which should work (it was working in the past), i.e., use the include shortcode and make sure to include `---` yaml header delimiters in the params file.

````
---
title: "Test"
format: html
---

{{< include _params.qmd >}}

# Test

```{r}
print(params$p1)
print(params$p2)
print(params$p3)
```
````

_params.qmd

```md
---
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
---
```

**Reply by GScottAlder** [reply]
That didn't quite work, since it creates two separate YML front matters
from the intermediate markdown file:
```
---
title: "Test"
format: html
---

---
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
---
```

However, if I take out the frontmatter entirely from test.qmd, then it works exactly how I want it to. The downside now is that I have to inherit the entire front matter in every .qmd file (I would prefer to only inherit the params), but that's a much better workaround

test.qmd:
```
{{< include _params.qmd >}}

# Test 

` ` ` {r}
print(params$p1)
print(params$p2)
print(params$p3)
` ` `
```

_params.qmd:
```
---
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
---
```

**Reply by mcanouil** [reply]
Wait, why did you say it did not work because of what you saw in an intermediate file?
How about the final file?


If this does not work this means there where some changes somewhere in how frontmatter in included files is handle since v1.0.

**Reply by cscheid** [reply]
To expand on what @mcanouil is saying, multiple front-matter entries on a file are allowed by Pandoc; that should not be the cause of the feature not working. (It's possible that something _else_ is causing the problem, still. But multiple front-matter entries are not it)

**Reply by GScottAlder** [reply]
So the following should be a valid *.qmd file?

```
---
title: "test"
format: html
---

---
params:
  p1: 1.0
  p2: "A"
  p3: TRUE
---

# Test 

` ` ` {r}
print(params$p1)
print(params$p2)
print(params$p3)
` ` `

```

When I try to render, I get the error "Error in print(params$p1) : object 'params' not found"

**Reply by mcanouil** [reply]
Yes.

This is the same

```
---
title: "Test"
format: html
---

---
author: me
---
```

as

```
---
title: "Test"
format: html
author: me
---
```

The above works, but for some reason, `params` is not recognised thus is not usable inside code cells.

@cscheid Might be best to open a new issue with the behaviour below unless you think it's somehow related. What's your thought?

Reproducible example:

1. Working as intended with the meta shortcode.

   ````
   ---
   title: "Test"
   format: html
   ---
 
   ---
   params:
     p1: 1.0
   ---
 
   {{< meta params.p1 >}}
   ````

2. Not working as intended within code cells.

   ````
   ---
   title: "Test"
   format: html
   ---
   
   ---
   params:
     p1: 1.0
   ---
   
   ```{r}
   params[["p1"]]
   ```
   ````

---

@GScottAlder Please don't insert spaces, use more backticks to encapsulate code with code cells.
For example, I used five backticks to show you what you should write when including code cells.

`````
````
---
title: "Test"
format: html
---

{{< include _params.qmd >}}

# Test

```{r}
print(params$p1)
print(params$p2)
print(params$p3)
```
````
`````

**Reply by msberends** [reply]
This does not answer the question - it focuses on `params`, while `metadata-files` can/should be used to inherit other properties, such as `author`. I can't seem to run R code that is inherited from the YAML file, and I believe that is in essence what this OP is also asking for.

**Reply by mcanouil** [reply]
Please open a new question, in this case because the main question was about an outdated version of Quarto and as you can easily see, no answer was marked as "the answer".
Also, most of the thread addressed another question.

So yes it does not answer the question but nobody said it did.

I am closing this thread as clearly outdated.
